### PR TITLE
[Fix #6893] Handle implicit rescue correctly in Naming/RescuedExceptionsVariableName

### LINKED ISSUE #6893: Naming/RescuedExceptionsVariableName invalid report
When I run rubocop __0.67.2__ with following code:
```ruby
handled = [StandardError]

begin
  raise StandardError
rescue *handled
  false
end
```

I get the following error:

```
test.rb:5:8: C: Naming/RescuedExceptionsVariableName: Use e instead of *handled.
rescue *handled
       ^^^^^^^^
```
--------

## Expected behavior

I should not get this offense.

## Actual behavior

I do get the offense.

## Steps to reproduce the problem

- Create a file with the code above
- Run Rubocop  on it

## RuboCop version

```
$ [bundle exec] rubocop -V
0.67.2 (using Parser 2.6.2.1, running on ruby 2.6.2 x86_64-linux)
```

### LINKED ISSUE #6907: Naming/RescuedExceptionsVariableName does not handle method call instead of an error class name
When I run rubocop `0.67.2` with following code:

```ruby
begin
  obj = ActiveSupport::JSON.decode(some_string)
rescue ActiveSupport::JSON.parse_error
  Rails.logger.warn("Attempted to decode invalid JSON: #{some_string}")
end
```

I get the following error:

```
C: Naming/RescuedExceptionsVariableName: Use e instead of ActiveSupport::JSON.parse_error.
    rescue ActiveSupport::JSON.parse_error
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

--------

## Expected behavior

I should not get this offense.

## Actual behavior

I do get the offense.

## Steps to reproduce the problem

* Create a file with the code above
* Run Rubocop on it

## RuboCop version

```
$ [bundle exec] rubocop -V
0.67.2 (using Parser 2.6.2.1, running on ruby 2.6.1 x86_64-darwin18)
```

### LINKED ISSUE #6912: RescuedExceptionsVariableName don't understand methods-as-rescue
In Ruby you can call a method in the rescue statement, and have it return an exception.

```ruby
module Errors
  def self.errors_matching_code(code)
    m = Module.new
    (class << m; self; end).instance_eval do
      define_method(:===, ->(e) { e.respond_to?(:code?) && e.code?(code) })
    end
    m
  end
end

begin
    ...
rescue Errors.errors_matching_code(1)
    ...
end
```

Rubocop 0.67.2 is currently unable to properly handle this case.

```
Naming/RescuedExceptionsVariableName: Use e instead of Errors.errors_matching_code(ERROR_CODE_SYSTEM_FAILEDUPDATE).
  rescue Errors.errors_matching_code(ERROR_CODE_SYSTEM_FAILEDUPDATE)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by pocke**
Fix #6893
Fix #6907
Fix #6912
Related #6908 Thanks @anthony-robin ! I've overlooked the pull request when I was implementing this pull request.

Problem
===

This cop has many issues. The cause is wrong handling for implicit rescue.

This cop handle `rescue e` as implicit rescue. It means the cop treats the code as an assignment to `e`. But it is not an assignment. It catches `e` class's error.
The cop should handle `rescue => e` as an implicit rescue.

Note
===

This pull request will remove `@exception_name` instance variable.
An instance of a cop is allocated each file. So the instance variable will be overridden if other `rescue` clauses exist.
So we need to avoid using the instance variable.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by robotdana** [comment]
This seems more correct than #6908

File: spec/rubocop/cop/naming/rescued_exceptions_variable_name_spec.rb [hunk 1]
```diff
@@ -113,6 +113,100 @@
...
+            rescue my_exception
+              # do something
+            end
+          RUBY
+        end
+      end
+
+      context 'with method as `Exception`' do
+        it 'does not register an offense without variable name' do
+          expect_no_offenses(<<-RUBY.strip_indent)
```
Comment (koic) [inline spec/rubocop/cop/naming/rescued_exceptions_variable_name_spec.rb]: Could you replace `<<-RUBY.strip_indent` with `<<~RUBY` in this PR?

Comment (pocke) [reply spec/rubocop/cop/naming/rescued_exceptions_variable_name_spec.rb]: I've replaced `strip_indent` with `<<~`. Thanks!

**Comment by bbatsov** [comment]
The changes look good, but your branch has to be rebased on top of the current `master` branch due to merge conflicts.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**